dashboard/app/models/pd/payment/payment_calculator_base.rb

```diff
@@ -0,0 +1,164 @@
+module Pd::Payment
+  SessionAttendanceSummary = Struct.new :hours, :teacher_ids
+  TeacherAttendanceTotal = Struct.new :days, :hours do
+    def add_session(hours)
+      self.days += 1
+      self.hours += hours
+    end
+  end
+
+  # Base class for Pd::Workshop payment. It calculates workshop duration, attendance numbers,
+  # and other general fields relevant to payment, and populates a #WorkshopPayment
+  # The class is a singleton, and should be used via #.instance.calculate(workshop)
+  # Derived classes can override to provide custom logic.
+  class PaymentCalculatorBase
+    include Singleton
+
+    DATE_FORMAT = '%m/%d/%Y'
+
+    # Calculates a WorkshopPayment from a workshop
+    # @param workshop [Pd::Workshop]
+    # @return [WorkshopPayment] the calculated payment.
+    def calculate(workshop)
+      raise 'Workshop required.' unless workshop
+      raise 'Workshop must be ended.' unless workshop.ended_at
+
+      WorkshopPayment.new.tap do |payment|
+        payment.workshop = workshop
+        payment.calculator_class = self.class
+        payment.pay_period = get_pay_period workshop
+
+        payment.num_days = workshop.effective_num_days
+        payment.num_hours = workshop.effective_num_hours
+        payment.min_attendance_days = workshop.min_attendance_days
+
+        payment.session_attendance_summaries = get_session_attendance_summaries workshop
+        payment.raw_teacher_attendance = calculate_raw_teacher_attendance payment.session_attendance_summaries
+        payment.adjusted_teacher_attendance = adjust_teacher_attendance(
+          payment.raw_teacher_attendance,
+          payment.min_attendance_days,
+          payment.num_days,
+          payment.num_hours
+        )
+
+        payment.total_teacher_attendance_days = payment.adjusted_teacher_attendance.values.map(&:days).reduce(0, :+)
+
+        payment.qualified = qualified? payment.total_teacher_attendance_days
+        payment.payment_type = get_payment_type workshop
+        payment.payment_amounts = calculate_payment_amounts payment
```

Comment:
I don't think we have a convention for this. I was going for one param per line for readability:

```ruby
payment.adjusted_teacher_attendance = adjust_teacher_attendance(
  payment.raw_teacher_attendance,
  payment.min_attendance_days,
  payment.num_days,
  payment.num_hours
)
```

Without parens it would look something like this (which I find to be less readable):
```ruby
payment.adjusted_teacher_attendance = adjust_teacher_attendance payment.raw_teacher_attendance,
  payment.min_attendance_days, 
  payment.num_days, 
  payment.num_hours
```
Preference?